pkg/cloudinit/cloudinit.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2022 SUSE.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudinit
+
+import (
+	"bytes"
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	bootstrapv1 "github.com/rancher-sandbox/cluster-api-provider-rke2/bootstrap/api/v1alpha1"
+)
+
+var (
+	defaultTemplateFuncMap = template.FuncMap{
+		"Indent": templateYAMLIndent,
+	}
+)
+
+func templateYAMLIndent(i int, input string) string {
+	split := strings.Split(input, "\n")
+	ident := "\n" + strings.Repeat(" ", i)
+	return strings.Repeat(" ", i) + strings.Join(split, ident)
+}
+
+const (
+	cloudConfigHeader = `## template: jinja
+#cloud-config
+`
+
+	filesTemplate = `{{ define "files" -}}
```

Comment:
I wonder if we should use yaml marshalling instead of using go templating.

Reply:
Although with yaml marshalling we'd need to add `#cloud-config` somehow

Reply:
Since Kubeadm is using this same approach, I would avoid, at least for now, to modify that. I can add an issue for investigating that. There is a Juju (Canonical) repo with something around [marshalling cloud-init](https://github.com/juju/juju/tree/develop/cloudconfig/cloudinit): 

It looks like a part of the challenge comes from the fact that cloud-init is not standardized and Opensuse, Ubuntu and Centos use different schemas.